File: workspace/src/my_py_pkg/my_py_pkg/minimal_subscriber.py
import rclpy
from rclpy.executors import ExternalShutdownException
from rclpy.node import Node
from example_interfaces.msg import String

class MinimalSubscriber(Node):
    """ Subscriber example that prints messages to the console """

    def __init__(self):
        super().__init__("minimal_subscriber")

        self._subscription = self.create_subscription(String, "my_topic", self._listener_callback, 10)

    def _listener_callback(self, msg: String):
        """ Prints message to the console """

        self.get_logger().info(f"Received message: {msg.data}")

def main(args = None):
    """ Main entrypoint """

    node = None
    try:
        rclpy.init()
        node = MinimalSubscriber()
        rclpy.spin(node)
    except (KeyboardInterrupt, ExternalShutdownException):
        pass
    finally:
        if node is not None:
            node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()

if __name__ == "__main__":
    main()

File: workspace/src/my_cpp_pkg/src/minimal_subscriber.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/msg/string.hpp"

class MinimalSubscriber : public rclcpp::Node 
{

public:

  MinimalSubscriber() : Node("minimal_subscriber")
  {
    subscriber_ = this->create_subscription<example_interfaces::msg::String>(
      "my_topic",
      10,
      std::bind(
        &MinimalSubscriber::listener_callback,
        this,
        std::placeholders::_1
      )
    );
  }

private:
  
  rclcpp::Subscription<example_interfaces::msg::String>::SharedPtr subscriber_;

  void listener_callback(const example_interfaces::msg::String& msg)
  {
    RCLCPP_INFO(this->get_logger(), "Received message: '%s'", msg.data.c_str());
  }

};

int main(int argc, char* argv[])
{
  rclcpp::init(argc, argv);

  auto node = std::make_shared<MinimalSubscriber>();
  rclcpp::spin(node);

  rclcpp::shutdown();

  return 0;
}